cli/src/api/mod.rs

```diff
@@ -415,33 +436,28 @@ impl PhylumApi {
     pub async fn get_project_history(
         &self,
         project_name: &str,
-        group_name: Option<&str>,
+        group: Option<Group>,
     ) -> Result<Vec<HistoryJob>> {
-        let project_id = self.get_project_id(project_name, group_name).await?.to_string();
-
-        let url = match group_name {
-            Some(group_name) => endpoints::get_group_project_history(
-                &self.config.connection.uri,
-                &project_id,
-                group_name,
-            )?,
-            None => endpoints::get_project_history(&self.config.connection.uri, &project_id)?,
-        };
-
+        let project_id = self.get_project_id(project_name, group).await?.to_string();
+        let url = endpoints::get_project_history(&self.config.connection.uri, &project_id)?;
         self.get(url).await
     }
 
     /// Resolve a Project Name to a Project ID
     pub async fn get_project_id(
         &self,
         project_name: &str,
-        group_name: Option<&str>,
+        group: Option<Group>,
     ) -> Result<ProjectId> {
-        let projects = self.get_projects(group_name, Some(project_name)).await?;
+        let (org, group, combined_format) = group.as_ref().map_or((None, None, None), |group| {
+            (group.org(), Some(group.name()), Some(group.combined_format().into_owned()))
+        });
```

Comment:
It seems like it would be easier to understand if we could write

```rust
let org = group.org();
let group = group.group(); // this naming isn't right
let group_path = group.combined_format();
```

The group parameter contains the organization-qualified group name (the group path?), but it's just called "group" and then another variable is defined named "group" which contains just the unqualified name of the group. It seems like the second "group" should be "group_name", but in the other functions "group_name" contains the qualified group name, which is here called "combined_format" instead.

Reply:
Your suggestion doesn't work, it's just an entirely different thing and does not behave the same way as mine.

Having `group_name` over `group` doesn't really improve things. It's just redundant.

Reply:
It would have been redundant before but I don't think it's redundant now. We're moving from a model where a group is identified by its name to a model where a group is identified by the name of the organization and the name of the group within the organization, which is leading to this problem where "group" can refer to the group or "group" can refer to the unqualified name of the group, which is just an attribute of the group and not enough to identify the group.

I think this is further complicated by the design of the CLI where the organization is being implied in contrast to the backend that the CLI uses where the organization is explicit. On the user-facing side, "group" means the unqualified name of the group within the configured organization, but then in the CLI command handler the unqualified name is upgraded into a qualified name, and from then on "group" _sometimes_ means the qualified group name. In this code we're decomposing the qualified "group" into its parts, and we're replacing `group: QualifiedGroup` with `org: GroupQualifier, group: UnqualifiedGroup`. That's the name overloading problem I have with other parts of this PR.

In this case though, I was more worried about `let (a, b, c) = o.as_ref().map_or((None, None, None), |v| (v.a(), Some(v.b()), Some(v.c().into_owned())));` We could write `let (a, b, c) = o.as_ref().map(|v| (v.a(), Some(v.b()), Some(v.c().into_owned()))).unwrap_or_default();`, but I don't think this is really much better. We could write `let (a, b, c) = match &o { Some(v) => (v.a(), Some(v.b()), Some(v.c().into_owned())), None => (None, None, None) }`. Originally I was thinking it would be nice if the parameter wasn't `Option<Group>` or if `Option<Group>` implemented a trait such that we could obtain information about the group or lack of group without mapping (eg `let org = group.org()`), but I think really what would be best here is if we didn't decompose the qualified group name at all: `let combined_format = group.map(|g| g.combined_format().into_owned());` here and then `let projects = self.get_projects(group, Some(project_name)).await?` later. It simplifies the code here in the caller and also slightly simplifies the code in the callee as well.

I think the `into_owned()` can be avoided by using `project.name == project_name && project.group_name.as_deref() == combined_format.as_deref())` for the filter.